File: app/src/main/java/com/pitchai/flickrrecentphotobrowser/presentation/fragments/MainGridFragmentList.java
package com.pitchai.flickrrecentphotobrowser.presentation.fragments;


import android.app.Activity;
import android.app.Fragment;
import android.content.Context;
import android.os.Bundle;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.pitchai.flickrrecentphotobrowser.AndroidApplication;
import com.pitchai.flickrrecentphotobrowser.R;
import com.pitchai.flickrrecentphotobrowser.RecyclerItemClickListener;
import com.pitchai.flickrrecentphotobrowser.data.photoinfo.Photo;
import com.pitchai.flickrrecentphotobrowser.presentation.adapter.RecyclerPhotoAdapter;
import com.pitchai.flickrrecentphotobrowser.presentation.internal.di.modules
        .MainGridListFragmentModule;
import com.pitchai.flickrrecentphotobrowser.presentation.presenter.GridPhotoListPresenterImp;
import com.pitchai.flickrrecentphotobrowser.presentation.view.GridPhotoListView;
import com.pitchai.flickrrecentphotobrowser.utils.Utils;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;


/**
 * A simple {@link Fragment} subclass.
 * Activities that contain this fragment must implement the
 * {@link MainGridFragmentList.OnFragmentInteractionListener} interface
 * to handle interaction events.
 * Use the {@link MainGridFragmentList#newInstance} factory method to
 * create an instance of this fragment.
 */
public class MainGridFragmentList extends BaseFragment implements GridPhotoListView {

    private static final String TAG = MainGridFragmentList.class.getSimpleName();
    public static int mColWidth;
    private OnFragmentInteractionListener mListener;
    SwipeRefreshLayout mSwipeRefresh;
    protected MenuItem mRefreshMenuItem;
    private RecyclerView mRecyclerView;
    private int scrollPreviousTotal = 0;
    private boolean loading = true;
    @Inject
    public RecyclerPhotoAdapter mRecyclerPhotoAdapter;
    @Inject
    public GridPhotoListPresenterImp gridPhotoListPresenterImp;
    ArrayList<Photo> mPhotoList;
    int mCurrentPageNumber;
    ProgressBar mProgressBar;
    //UsersAdapter usersAdapter;

    public MainGridFragmentList() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @return A new instance of fragment MainGridFragmentList.
     */
    public static MainGridFragmentList newInstance() {
        MainGridFragmentList fragment = new MainGridFragmentList();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
        setHasOptionsMenu(true);
        mPhotoList = new ArrayList<>();
        AndroidApplication.getAndroidApplication(this.getActivity())
                .getApplicationComponent()
                .plus(new MainGridListFragmentModule(this))
                .inject(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        View view = inflater.inflate(R.layout.fragment_main_grid, container, false);
        initializeView(view);
        // Inflate the layout for this fragment
        return view;
    }


    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener) {
            mListener = (OnFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onAttach(Activity context) {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener) {
            mListener = (OnFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    public void onDestroy() {
        super.onDestroy();
        gridPhotoListPresenterImp.destroy();
    }

    /**
     * Initialize Views
     */
    private void initializeView(View view) {

        mProgressBar = (ProgressBar) view.findViewById(R.id.activity_progressbar);
        mSwipeRefresh = (SwipeRefreshLayout) view.findViewById(R.id.swipeRefresh);
        // usersAdapter = new UsersAdapter(getActivity().getApplicationContext());
        mRecyclerView = (RecyclerView) view.findViewById(R.id.photo_recycler_view);
        /*mRecyclerView.setLayoutManager(new UsersLayoutManager(getActivity()
        .getApplicationContext()));
        mRecyclerView.setAdapter(usersAdapter);*/
        mRecyclerView.setHasFixedSize(true);
        mColWidth = Utils.calculateWidth(getActivity());
        final GridLayoutManager layoutManager = new GridLayoutManager(getActivity()
                .getApplicationContext(), Utils.calculateWidthScale(getActivity()));
        mRecyclerView.setLayoutManager(layoutManager);
        mRecyclerView.setAdapter(mRecyclerPhotoAdapter);

        mRecyclerView.addOnItemTouchListener(new RecyclerItemClickListener(getActivity()
                .getApplicationContext(), new RecyclerItemClickListener.OnItemClickListener() {
            @Override
            public void onItemClick(View view, int position) {
                Log.d(TAG, "position:" + position);
                mListener.onPhotoSelect(mPhotoList, position);
            }
        }));
        mRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int x, int y) {
                super.onScrolled(recyclerView, x, y);
                int totalItemCount = layoutManager.getItemCount();
                int firstVisibleItem = layoutManager.findFirstVisibleItemPosition();
                int visibleItemCount = recyclerView.getChildCount();
                if (loading) {
                    if (totalItemCount > scrollPreviousTotal) {
                        loading = false;
                        scrollPreviousTotal = totalItemCount;
                    }
                }
                if (!loading && (totalItemCount - visibleItemCount) <= (firstVisibleItem +
                        visibleItemCount)) {
                    //Load more page
                    loadMore();
                    loading = true;
                }
            }
        });

        mSwipeRefresh.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                refresh();
            }
        });
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        gridPhotoListPresenterImp.setGridPhotoListView(this);
        if (mPhotoList.size() == 0) {
            gridPhotoListPresenterImp.execute(1);
        } else {
            mRecyclerPhotoAdapter.updateList(mPhotoList);
        }
    }

    @Override
    public void renderPhotoList(List<Photo> photoList, int currentPage) {
        if (mSwipeRefresh.isRefreshing()) {
            stopRefreshProgress();
        }
        mCurrentPageNumber = currentPage;
        mPhotoList.addAll(photoList);
        //Update the adapter
        mRecyclerPhotoAdapter.updateList(photoList);
    }


    @Override
    public void showLoading() {
        mProgressBar.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideLoading() {
        mProgressBar.setVisibility(View.GONE);
    }

    @Override
    public void showError(String message) {
        Toast.makeText(getActivity(), message, Toast.LENGTH_SHORT).show();
    }

    @Override
    public Context context() {
        return null;
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     * <p>
     * See the Android Training lesson <a href=
     * "http://developer.android.com/training/basics/fragments/communicating.html"
     * >Communicating with Other Fragments</a> for more information.
     */
    public interface OnFragmentInteractionListener {
        void onPhotoSelect(ArrayList<Photo> photoList, int position);
    }

    private void loadMore() {
        gridPhotoListPresenterImp.execute(mCurrentPageNumber + 1);
    }


    private void refresh() {
        scrollPreviousTotal = 0;
        mPhotoList.clear();
        mRecyclerPhotoAdapter.clear();
        gridPhotoListPresenterImp.execute(1);
    }

    /**
     * Remove the refresh views on complete
     */
    private void stopRefreshProgress() {
        if (mRefreshMenuItem != null) {
            mRecyclerView.getLayoutManager().scrollToPosition(0);
            mRefreshMenuItem.setActionView(null);
        }
        if (mSwipeRefresh.isRefreshing()) {
            mSwipeRefresh.setRefreshing(false);
        }
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.main_menu, menu);
        mRefreshMenuItem = menu.findItem(R.id.refresh);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {

            case R.id.refresh:
                mRefreshMenuItem.setActionView(R.layout.actionbar_progress);
                refresh();
                return true;
        }
        return false;
    }
}

File: app/src/main/java/com/pitchai/flickrrecentphotobrowser/presentation/fragments/MainGridFragmentList.kt
package com.pitchai.flickrrecentphotobrowser.presentation.fragments

import android.content.Context
import android.os.Bundle
import android.view.*
import android.widget.Toast
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.paging.LoadState
import androidx.recyclerview.widget.GridLayoutManager
import com.pitchai.flickrrecentphotobrowser.AndroidApplication
import com.pitchai.flickrrecentphotobrowser.R
import com.pitchai.flickrrecentphotobrowser.databinding.FragmentMainGridBinding
import com.pitchai.flickrrecentphotobrowser.di.MainGridListFragmentModule
import com.pitchai.flickrrecentphotobrowser.presentation.ViewModel.PhotoListViewModel
import com.pitchai.flickrrecentphotobrowser.presentation.adapter.FlickrLoadStateAdapter
import com.pitchai.flickrrecentphotobrowser.presentation.adapter.PhotoClickListener
import com.pitchai.flickrrecentphotobrowser.presentation.adapter.RecyclerPhotoListAdapter
import com.pitchai.flickrrecentphotobrowser.presentation.view.LoadDataView
import com.pitchai.flickrrecentphotobrowser.utils.Utils
import kotlinx.coroutines.InternalCoroutinesApi
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * A simple [Fragment] subclass.
 * Activities that contain this fragment must implement the
 * [MainGridFragmentList.OnFragmentInteractionListener] interface
 * to handle interaction events.
 * Use the [MainGridFragmentList.newInstance] factory method to
 * create an instance of this fragment.
 */
@InternalCoroutinesApi
class MainGridFragmentList : BaseFragment(), LoadDataView, PhotoClickListener {
    @Inject
    lateinit var mRecyclerPhotoAdapter: RecyclerPhotoListAdapter

    @Inject
    lateinit var photoListViewModel: PhotoListViewModel

    private var mListener: OnFragmentInteractionListener? = null
    protected var mRefreshMenuItem: MenuItem? = null
    private lateinit var binding: FragmentMainGridBinding


    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        binding = FragmentMainGridBinding.inflate(inflater, container, false)
        val view = binding.root
        setHasOptionsMenu(true)
        initializeView(view)
        // Inflate the layout for this fragment
        return view
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        AndroidApplication.getAndroidApplication(this.activity).applicationComponent
            ?.plus(MainGridListFragmentModule(this))
            ?.inject(this)
        mListener = if (context is OnFragmentInteractionListener) {
            context
        } else {
            throw RuntimeException(
                context.toString()
                        + " must implement OnFragmentInteractionListener"
            )
        }
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }

    /**
     * Initialize Views
     */
    private fun initializeView(view: View) {

        binding.photoRecyclerView.setHasFixedSize(true)
        val mColWidth = Utils.calculateWidth(requireActivity())
        mRecyclerPhotoAdapter.columnWidth = mColWidth
        val layoutManager = GridLayoutManager(
            activity
                ?.applicationContext, Utils.calculateWidthScale(requireActivity())
        )
        binding.photoRecyclerView.layoutManager = layoutManager
        binding.photoRecyclerView.adapter = mRecyclerPhotoAdapter.withLoadStateHeaderAndFooter(
            header = FlickrLoadStateAdapter { mRecyclerPhotoAdapter.retry() },
            footer = FlickrLoadStateAdapter { mRecyclerPhotoAdapter.retry() }
        )

        mRecyclerPhotoAdapter.addLoadStateListener { loadState ->

            // show empty list
            val isListEmpty =
                loadState.refresh is LoadState.NotLoading && mRecyclerPhotoAdapter.itemCount == 0
            showEmptyList(isListEmpty)

            // Only show the list if refresh succeeds.
            binding.photoRecyclerView.isVisible =
                loadState.mediator?.refresh is LoadState.NotLoading
            // Show loading spinner during initial load or refresh.
            binding.progressBar.isVisible = loadState.mediator?.refresh is LoadState.Loading
            // Show the retry state if initial load or refresh fails.
            //binding.retryButton.isVisible = loadState.mediator?.refresh is LoadState.Error
            // Toast on any error, regardless of whether it came from RemoteMediator or PagingSource
            val errorState = loadState.source.append as? LoadState.Error
                ?: loadState.source.prepend as? LoadState.Error
                ?: loadState.append as? LoadState.Error
                ?: loadState.prepend as? LoadState.Error
            errorState?.let {
                showError("\uD83D\uDE28 Wooops ${it.error}")
            }
        }
        // Scroll to top when the list is refreshed from network.
        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.lifecycle.repeatOnLifecycle(Lifecycle.State.STARTED) {
//                mRecyclerPhotoAdapter.loadStateFlow
//                    // Only emit when REFRESH LoadState for RemoteMediator changes.
//                    .distinctUntilChangedBy {
//                        it.refresh
//                    }
//                    // Only react to cases where Remote REFRESH completes i.e., NotLoading.
//                    .filter {
//                        it.refresh is LoadState.NotLoading
//                    }
//                    .collect {
//                        mRecyclerView?.scrollToPosition(0)
//                    }
                photoListViewModel.getPhotoList().collectLatest {
                    stopRefreshProgress()
                    mRecyclerPhotoAdapter.submitData(it)
                }
            }
        }
    }

    override fun showLoading() {
        binding.progressBar.visibility = View.VISIBLE
    }

    override fun hideLoading() {
        binding.progressBar.visibility = View.GONE
    }

    override fun showError(message: String?) {
        Toast.makeText(activity, message, Toast.LENGTH_SHORT).show()
    }

    private fun showEmptyList(show: Boolean) {
        if (show) {
            binding.emptyList.visibility = View.VISIBLE
            binding.photoRecyclerView.visibility = View.GONE
        } else {
            binding.emptyList.visibility = View.GONE
            binding.photoRecyclerView.visibility = View.VISIBLE
        }
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     *
     *
     */
    interface OnFragmentInteractionListener {
        fun onPhotoSelect(position: Int)
    }

    private fun refresh() {
        mRecyclerPhotoAdapter.refresh()
    }

    /**
     * Remove the refresh views on complete
     */
    private fun stopRefreshProgress() {
        if (mRefreshMenuItem != null) {
            //mRecyclerView?.layoutManager?.scrollToPosition(0)
            mRefreshMenuItem?.setActionView(null)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.main_menu, menu)
        mRefreshMenuItem = menu.findItem(R.id.refresh)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.refresh -> {
                mRefreshMenuItem?.setActionView(R.layout.actionbar_progress)
                refresh()
                return true
            }
        }
        return false
    }

    companion object {
        private val TAG = MainGridFragmentList::class.java.simpleName

        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @return A new instance of fragment MainGridFragmentList.
         */
        fun newInstance(): Fragment {
            val fragment = MainGridFragmentList()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }
    }

    override fun onItemClick(position: Int) {
        mListener?.onPhotoSelect(position)
    }
}